Source/WebDriver/Session.cpp

```diff
@@ -3282,21 +3283,119 @@ void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
 }
 
-void Session::enableGlobalEvent(const String& eventName)
+void Session::subscribeForEvents(const Vector<String>& events, Vector<String>&& browsingContextIDs, Vector<String>&& userContextIDs, Function<void(CommandResult&&)>&& completionHandler)
 {
-    m_globalEventSet.add(eventName);
+    // FIXME: Process/validate list of event names (e.g. expanding if given only the module name)
+    // https://bugs.webkit.org/show_bug.cgi?id=291371
+    auto subscriptionID = WTF::createVersion4UUIDString();
+
+    for (const auto& event : events) {
+        auto addResult = m_eventSubscriptionCount.add(AtomString { event }, 1);
+        if (!addResult.isNewEntry)
+            addResult.iterator->value++;
+    }
+
+    Vector<AtomString> atomEventNames;
+    for (const auto& event : events)
+        atomEventNames.append(AtomString { event });
+
+    m_eventSubscriptions.add(subscriptionID, EventSubscription { subscriptionID, WTFMove(atomEventNames), WTFMove(browsingContextIDs), WTFMove(userContextIDs) });
+    completionHandler(CommandResult::success(JSON::Value::create(subscriptionID)));
 }
 
-void Session::disableGlobalEvent(const String& eventName)
+void Session::unsubscribeByIDs(const Vector<EventSubscriptionID>& subscriptionIDs, Function<void(CommandResult&&)>&& completionHandler)
 {
-    m_globalEventSet.remove(eventName);
+    for (const auto& id : subscriptionIDs) {
+        if (!m_eventSubscriptions.contains(id)) {
+            completionHandler(CommandResult::fail(CommandResult::ErrorCode::InvalidArgument, "At least one subscription id is unknown"_s));
+            return;
+        }
+    }
+
+    for (const auto& id : subscriptionIDs) {
+        const auto& subscription = m_eventSubscriptions.get(id);
+
+        for (const auto& event : subscription.events) {
+            auto removeResult = m_eventSubscriptionCount.find(event);
+            if (removeResult != m_eventSubscriptionCount.end()) {
+                if (!(--removeResult->value))
+                    m_eventSubscriptionCount.remove(event);
+            }
+        }
+        m_eventSubscriptions.remove(id);
+    }
+    completionHandler(CommandResult::success());
+}
+
+void Session::unsubscribeByEventName(const Vector<String>& eventNames, Function<void(CommandResult&&)>&& completionHandler)
+{
+    HashMap<String, EventSubscription> subscriptionsToKeep;
+    HashSet<String> matchedEvents;
+    // TODO Process/validate list of event names (e.g. expanding if given only the module name)
+    // https://bugs.webkit.org/show_bug.cgi?id=291371
+    for (const auto& eventName : eventNames) {
+        auto atomEventName = AtomString { eventName };
+        for (const auto& subscription : m_eventSubscriptions) {
+            if (!subscription.value.events.contains(atomEventName)) {
+                subscriptionsToKeep.add(subscription.value.id, subscription.value);
+                continue;
+            }
+
+            // FIXME Add support to subscribe to specific browsing contexts
```

Comment:
Style nit: I usually write FIXME with : (FIXME:). And prefer TODOs removed or converted to FIXMEs.

Reply:
Oops, I fixed the other one and missed this.